segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -279,10 +268,31 @@ private void handleAppendResult(final Append append, Throwable exception) {
                 if (statsRecorder != null) {
                     statsRecorder.record(append.getSegment(), append.getDataLength(), append.getEventCount());
                 }
-                connection.send(new DataAppended(append.getWriterId(), append.getEventNumber(), previousEventNumber));
+                final DataAppended dataAppendedAck = new DataAppended(append.getWriterId(), append.getEventNumber(),
+                        previousEventNumber);
+                log.trace("Sending DataAppended : {}", dataAppendedAck);
+                connection.send(dataAppendedAck);
                 DYNAMIC_LOGGER.incCounterValue(nameFromSegment(SEGMENT_WRITE_BYTES, append.getSegment()), append.getDataLength());
                 DYNAMIC_LOGGER.incCounterValue(nameFromSegment(SEGMENT_WRITE_EVENTS, append.getSegment()), append.getEventCount());
             }
+
+            /* Reply (DataAppended in case of success, else an error Reply based on exception) has been sent. Next,
+             *   - clear outstandingAppend to handle the next Append message.
+             *   - ensure latestEventNumbers and waitingAppends are updated.
+             */
+            synchronized (lock) {
```

Comment:
I might be missing something, but I'm not convinced this is going to work. My recollection of the problem is that a segment store thread invokes this method and calls `connect.send`. The actual send is executed by a netty thread and according to the logs I have seen, the send returns without having actually sent it. I understand that the delay is due to thread scheduling.

The acks get out of order because the netty threads can also invoke this method, in which case, the send can be executed immediately causing the order issue we are observing.

If this insight is incorrect, then I need some justification of why that's the case.

Reply:
`AppendProcessor#append` is invoked by a netty thread. This invokes `performNextWrite` which fetchs the append and writes it to the store (Async write). Once this write is complete it invokes `handleAppendResult()`. The step of writing on store is done by the segmentstore thread and most of the times `storeAppend(append).whenComplete(...) is executed on that thread.

Now `handleAppendResult()` internally sends the ack, clears `outstandingAppend` and then invokes `performNextWrite()`. PerformNextWrite will processing the next append only if  previous ack is sent, if there is a backlog then the segmentstore-store thread might be used for the next performNextWrite() invocation. Now, if there is a delay is sending ack and in the mean time a new Append is received `AppendProcessor#append and internally performNextWrite` is invoked on a different netty thread. (This used to cause a race in the older code, as the next append could cause the ack to be sent while older ack has not yet been sent.)